File: packages/Thoth.Json.Core/Object.fs
namespace Thoth.Json.Core

[<AutoOpen>]
module ObjectCodecComputationExpression =

    /// <summary>
    /// The fields declared so far by an <c>objectCodec</c> block.
    /// </summary>
    [<NoComparison>]
    [<NoEquality>]
    type ObjectCodecFieldSet<'t, 'u> =
        {
            Values: 't -> (string * IEncodable) list
            Decoder: Decoder<'t>
            Picker: 'u -> 't
        }

    [<RequireQualifiedAccess>]
    module internal ObjectCodecFieldSet =

        let zip
            (a: ObjectCodecFieldSet<'a, 'u>)
            (b: ObjectCodecFieldSet<'b, 'u>)
            : ObjectCodecFieldSet<'a * 'b, 'u>
            =
            {
                Values = fun (i, j) -> a.Values i @ b.Values j
                Decoder = Decode.map2 (fun i j -> i, j) a.Decoder b.Decoder
                Picker = fun u -> a.Picker u, b.Picker u
            }

        let complete
            (f: 't -> 'u)
            (m: ObjectCodecFieldSet<'t, 'u>)
            : Codec<'u>
            =
            let codec =
                Codec.create (fun t -> m.Values t |> Encode.object) m.Decoder

            Codec.map f m.Picker codec

    /// <summary>
    /// The builder behind <c>objectCodec</c>.
    /// </summary>
    type ObjectCodecBuilder internal () =
        member this.MergeSources
            (a: ObjectCodecFieldSet<'a, 'u>, b: ObjectCodecFieldSet<'b, 'u>)
            : ObjectCodecFieldSet<'a * 'b, 'u>
            =
            ObjectCodecFieldSet.zip a b

        member this.BindReturn
            (m: ObjectCodecFieldSet<'t, 'u>, f: 't -> 'u)
            : Codec<'u>
            =
            ObjectCodecFieldSet.complete f m

    /// <summary>
    /// Build a codec for a record, one <see cref="M:Thoth.Json.Core.Codec.field"/> per property.
    /// </summary>
    /// <example>
    /// <code lang="fsharp">
    /// let codec : Codec&lt;Point&gt; =
    ///     objectCodec {
    ///         let! x = Codec.field "x" _.X Codec.int
    ///         and! y = Codec.field "y" _.Y Codec.int
    ///
    ///         return { X = x; Y = y }
    ///     }
    /// </code>
    /// </example>
    let objectCodec = ObjectCodecBuilder()

    [<RequireQualifiedAccess>]
    module Codec =

        /// <summary>
        /// A required property of an <c>objectCodec</c> block: its name, how to read it out of the
        /// value, and the codec for it.
        /// </summary>
        let field
            (fieldName: string)
            (picker: 'u -> 't)
            (fieldCodec: Codec<'t>)
            : ObjectCodecFieldSet<'t, 'u>
            =
            {
                Values = fun i -> [ fieldName, fieldCodec.Encoder i ]
                Decoder = Decode.field fieldName fieldCodec.Decoder
                Picker = picker
            }

        /// <summary>
        /// An optional property of an <c>objectCodec</c> block. It is left out when the value is
        /// <c>None</c>, and its absence decodes back to <c>None</c>.
        /// </summary>
        let optional
            (fieldName: string)
            (picker: 'u -> 't option)
            (fieldCodec: Codec<'t>)
            : ObjectCodecFieldSet<'t option, 'u>
            =
            {
                Values =
                    function
                    | Some i -> [ fieldName, fieldCodec.Encoder i ]
                    | None -> []
                Decoder = Decode.optional fieldName fieldCodec.Decoder
                Picker = picker
            }

File: assets/live-example/precompiled/e667cbc66489906b/packages/Thoth.Json.Core/Object.js

import { Record } from "fable-library-js/Types.js";
import { record_type, lambda_type, list_type, tuple_type, class_type, string_type } from "fable-library-js/Reflection.js";
import { empty, singleton, append } from "fable-library-js/List.js";
import { optional, field, map2 } from "./Decode.js";
import { create, map } from "./Codec.js";
import { map as map_1 } from "fable-library-js/Seq.js";
import { value } from "fable-library-js/Option.js";

/**
 * The fields declared so far by an <c>objectCodec</c> block.
 */
export class ObjectCodecFieldSet$2 extends Record {
    constructor(Values, Decoder, Picker) {
        super();
        this.Values = Values;
        this.Decoder = Decoder;
        this.Picker = Picker;
    }
}

export function ObjectCodecFieldSet$2_$reflection(gen0, gen1) {
    return record_type("Thoth.Json.Core.ObjectCodecComputationExpression.ObjectCodecFieldSet`2", [gen0, gen1], ObjectCodecFieldSet$2, () => [["Values", lambda_type(gen0, list_type(tuple_type(string_type, class_type("Thoth.Json.Core.IEncodable"))))], ["Decoder", class_type("Thoth.Json.Core.Decoder`1", [gen0])], ["Picker", lambda_type(gen1, gen0)]]);
}

export function ObjectCodecFieldSet_zip(a, b) {
    return new ObjectCodecFieldSet$2((tupledArg) => append(a.Values(tupledArg[0]), b.Values(tupledArg[1])), map2((i_1, j_1) => [i_1, j_1], a.Decoder, b.Decoder), (u) => [a.Picker(u), b.Picker(u)]);
}

export function ObjectCodecFieldSet_complete(f, m) {
    return map(f, m.Picker, create((t) => {
        const values = m.Values(t);
        return {
            Encode(helpers) {
                const arg = map_1((tupledArg) => [tupledArg[0], tupledArg[1].Encode(helpers)], values);
                return helpers.encodeObject(arg);
            },
        };
    }, m.Decoder));
}

/**
 * The builder behind <c>objectCodec</c>.
 */
export class ObjectCodecBuilder {
    constructor() {
    }
}

export function ObjectCodecBuilder_$reflection() {
    return class_type("Thoth.Json.Core.ObjectCodecComputationExpression.ObjectCodecBuilder", undefined, ObjectCodecBuilder);
}

export function ObjectCodecBuilder_$ctor() {
    return new ObjectCodecBuilder();
}

export function ObjectCodecBuilder__MergeSources_5532C640(this$, a, b) {
    return ObjectCodecFieldSet_zip(a, b);
}

export function ObjectCodecBuilder__BindReturn_Z2BDEB988(this$, m, f) {
    return ObjectCodecFieldSet_complete(f, m);
}

export const objectCodec = ObjectCodecBuilder_$ctor();

/**
 * A required property of an <c>objectCodec</c> block: its name, how to read it out of the
 * value, and the codec for it.
 */
export function Codec_field(fieldName, picker, fieldCodec) {
    return new ObjectCodecFieldSet$2((i) => singleton([fieldName, fieldCodec.Encoder(i)]), field(fieldName, fieldCodec.Decoder), picker);
}

/**
 * An optional property of an <c>objectCodec</c> block. It is left out when the value is
 * <c>None</c>, and its absence decodes back to <c>None</c>.
 */
export function Codec_optional(fieldName, picker, fieldCodec) {
    return new ObjectCodecFieldSet$2((_arg) => ((_arg == null) ? empty() : singleton([fieldName, fieldCodec.Encoder(value(_arg))])), optional(fieldName, fieldCodec.Decoder), picker);
}

